### ISSUE TITLE
Question about implementation details on Cifar100 dataset.

### ISSUE THREAD

**Issue by mayorx** [YingZhangDUT/Deep-Mutual-Learning#3]
I tested Resnet-32 (both Independent and DML) on Cifar100 dataset, but I didn't get significant improvement using DML, was my experimental settings incorrect?

I followed your experimental settings, which is:
* learning rate
   - 0.1 (epoch < 60)
   - 0.01 (epoch < 120)
   - 0.001 (epoch < 180)
   - 0.0001 (otherwise)
* momentum 0.9
* batch-size 64
* epoch 200
* weight-decay 5e-4
* using SGD
    - set nesterov=True

my result of independent resnet-32 is (average of 6 results) 70.98% , and DML of resnet-32 and resnet-32 is 70.86% / 70.90% .

**Comment by mayorx** [comment]
more results are as follows:
* DML (Resnet-32 && Resnet-32):
    - 70.33/71.13
    - 70.58/71.45
    - 71.02/70.37
    - 71.06/70.65
    - 71.34/70.94
    - 70.84/70.84
* Independent
    - 70.98
    - 70.45
    - 71.35
    - 71.6
    - 70.56
    - 70.93

**Comment by Guo-Xiaoqing** [comment]
> more results are as follows:
> 
> * DML (Resnet-32 && Resnet-32):
>   
>   * 70.33/71.13
>   * 70.58/71.45
>   * 71.02/70.37
>   * 71.06/70.65
>   * 71.34/70.94
>   * 70.84/70.84
> * Independent
>   
>   * 70.98
>   * 70.45
>   * 71.35
>   * 71.6
>   * 70.56
>   * 70.93

I don't understand why accuracy on cifar100 dataset is always around 0.1 when I train, and the initial learning rate seems to be too large?  I also followed your experimental settings. Did you once meet this problem? In the paper, the predictions should be updated before computer the gradient and update theta2 in a mini-batch, but the code seems not to update predictions in a mini-batch, right?

**Comment by Guo-Xiaoqing** [comment]
> more results are as follows:
> 
> * DML (Resnet-32 && Resnet-32):
>   
>   * 70.33/71.13
>   * 70.58/71.45
>   * 71.02/70.37
>   * 71.06/70.65
>   * 71.34/70.94
>   * 70.84/70.84
> * Independent
>   
>   * 70.98
>   * 70.45
>   * 71.35
>   * 71.6
>   * 70.56
>   * 70.93

I don't understand why accuracy on cifar100 dataset is always around 0.1 when I trained, and the initial learning rate seems to be too large? I also followed your experimental settings. Did you once meet this problem? In the paper, the predictions should be updated before computering the gradient and updating theta2 in a mini-batch, but the code seems not to update predictions in a mini-batch, right?

**Comment by mayorx** [comment]
> > more results are as follows:
> > 
> > * DML (Resnet-32 && Resnet-32):
> >   
> >   * 70.33/71.13
> >   * 70.58/71.45
> >   * 71.02/70.37
> >   * 71.06/70.65
> >   * 71.34/70.94
> >   * 70.84/70.84
> > * Independent
> >   
> >   * 70.98
> >   * 70.45
> >   * 71.35
> >   * 71.6
> >   * 70.56
> >   * 70.93
> 
> I don't understand why accuracy on cifar100 dataset is always around 0.1 when I trained, and the initial learning rate seems to be too large? I also followed your experimental settings. Did you once meet this problem? In the paper, the predictions should be updated before computering the gradient and updating theta2 in a mini-batch, but the code seems not to update predictions in a mini-batch, right?

First question , I recommend you using a tested code first (for example, https://github.com/bearpaw/pytorch-classification), and then you can find out the difference between your model and the tested one.
Second question already been answered here: https://github.com/YingZhangDUT/Deep-Mutual-Learning/issues/2

**Comment by Guo-Xiaoqing** [comment]
Do you know why you didn't get significant improvement using DML now? I also wonder this problem.

**Comment by Guo-Xiaoqing** [comment]
> > > more results are as follows:
> > > 
> > > * DML (Resnet-32 && Resnet-32):
> > >   
> > >   * 70.33/71.13
> > >   * 70.58/71.45
> > >   * 71.02/70.37
> > >   * 71.06/70.65
> > >   * 71.34/70.94
> > >   * 70.84/70.84
> > > * Independent
> > >   
> > >   * 70.98
> > >   * 70.45
> > >   * 71.35
> > >   * 71.6
> > >   * 70.56
> > >   * 70.93
> > 
> > 
> > I don't understand why accuracy on cifar100 dataset is always around 0.1 when I trained, and the initial learning rate seems to be too large? I also followed your experimental settings. Did you once meet this problem? In the paper, the predictions should be updated before computering the gradient and updating theta2 in a mini-batch, but the code seems not to update predictions in a mini-batch, right?
> 
> First question , I recommend you using a tested code first (for example, https://github.com/bearpaw/pytorch-classification), and then you can find out the difference between your model and the tested one.
> Second question already been answered here: #2

Do you know why you didn't get significant improvement using DML now? I also wonder this problem.

**Comment by mayorx** [comment]
> > 
> > 
> > First question , I recommend you using a tested code first (for example, https://github.com/bearpaw/pytorch-classification), and then you can find out the difference between your model and the tested one.
> > Second question already been answered here: #2
> 
> Do you know why you didn't get significant improvement using DML now? I also wonder this problem.

Sorry, I don't know.. that's why this issue is still open

**Comment by justdoxiaolin** [comment]
Hello, I am a beginner of deep learning.I read the code you wrote about the market1501 dataset.
I tried the cifar100 dataset myself, but it didn't work. Can you share the code about the cifar100 dataset, thank you! Sincerely need your help. @mayorx @Guo-Xiaoqing

**Comment by mayorx** [comment]
> Hello, I am a beginner of deep learning.I read the code you wrote about the market1501 dataset.
> I tried the cifar100 dataset myself, but it didn't work. Can you share the code about the cifar100 dataset, thank you! Sincerely need your help. @mayorx @Guo-Xiaoqing

Sorry, I didn't keep the code.  What I did is just run a [cifar-classification model](https://github.com/bearpaw/pytorch-classification) and add the kl_loss, which you could refer to this repo:

https://github.com/YingZhangDUT/Deep-Mutual-Learning/blob/34a20583debe4e0dab1d9856db69bed278c5c011/train_models.py#L60-L67

https://github.com/YingZhangDUT/Deep-Mutual-Learning/blob/34a20583debe4e0dab1d9856db69bed278c5c011/train_models.py#L94

**Comment by justdoxiaolin** [comment]
thank you! I am still a little confused, can you write that code again? I need your help sincerely.@mayorx

**Comment by justdoxiaolin** [comment]
Because there are a lot of things you don't understand, can you rewrite the code? Although I know that it is very troublesome, I sincerely beg you can help me.
 @mayorx @Guo-Xiaoqing

**Comment by mayorx** [comment]
> Because there are a lot of things you don't understand, can you rewrite the code? Although I know that it is very troublesome, I sincerely beg you can help me.
> @mayorx @Guo-Xiaoqing

here is my [code](https://paste.ubuntu.com/p/cxPRRs4x6W/) , which is based on [cifar-classification model](https://github.com/junyuseu/pytorch-cifar-models).
I wrote it about 1 year ago.

**Comment by justdoxiaolin** [comment]
Thank you! Thanks for sharing! thank you very much.I will learn from you.




------------------ 原始邮件 ------------------
发件人: "Chen Liangyu"<notifications@github.com>;
发送时间: 2019年2月28日(星期四) 晚上9:07
收件人: "YingZhangDUT/Deep-Mutual-Learning"<Deep-Mutual-Learning@noreply.github.com>;
抄送: "千与千寻"<1875823709@qq.com>; "Comment"<comment@noreply.github.com>; 
主题: Re: [YingZhangDUT/Deep-Mutual-Learning] Question about implementationdetails on Cifar100 dataset. (#3)



 
Because there are a lot of things you don't understand, can you rewrite the code? Although I know that it is very troublesome, I sincerely beg you can help me.
 @mayorx @Guo-Xiaoqing
  
here is my code , which is based on cifar-classification model.
 I wrote it about 1 year ago.

**Comment by justdoxiaolin** [comment]
Thank you! Thanks for sharing! thank you very much.I will learn from you. @mayorx

**Comment by ljh9999** [comment]
How does this code change to run on the cifar data set?

**Comment by chxy95** [comment]
Have you found out the reason? 
I train it and the result shows there are a little promotion of performance.
However, when use two models to do DML, the promotion is not so large as the paper shows.
As my result, the two-model DML only have about 1.2% promotion.
But when use more models to train, the promotion is more obvious.
I have tried to use five models to train, and get about 2% promotion of performance.

**Comment by zyyupup** [comment]
I found  that i can't reach the accuracy 76.13 of mobilenet shown in the paper. And i also found that the results of WRN-28-10 in the  original paper is 80.75 while in DML is 78.69. Anyone  knows the reason?

**Comment by JiyueWang** [comment]
May I ask a question about the baseline accuracy. https://github.com/weiaicunzai/pytorch-cifar100
In the repo above the resnet34 accuracy is much higher 76.76% than this paper, do you know what is the difference. Thanks.
